smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -1611,25 +1634,35 @@ private void writeEventBody(GenerationContext context, StructureShape event) {
         TypeScriptWriter writer = context.getWriter();
         Model model = context.getModel();
         List<MemberShape> payloadMembers = event.getAllMembers().values().stream()
-                .filter(member -> member.hasTrait(EventPayloadTrait.class)).collect(Collectors.toList());
-        List<MemberShape> documentMembers = event.getAllMembers().values().stream()
-                .filter(member -> !member.hasTrait(EventHeaderTrait.class)
-                        && !member.hasTrait(EventPayloadTrait.class))
+                .filter(member -> member.hasTrait(EventPayloadTrait.class))
                 .collect(Collectors.toList());
-        if (!payloadMembers.isEmpty()) {
-            // Write event payload if exists. There is at most 1 payload member.
+        Shape payloadShape = payloadMembers.isEmpty()
+                        ? event // implicit payload
+                        : model.expectShape(payloadMembers.get(0).getTarget());
+        if (payloadShape instanceof BlobShape || payloadShape instanceof StringShape) {
+            // Since event itself must be a structure shape, so string or blob payload member must has eventPayload
+            // trait explicitly.
             MemberShape payloadMember = payloadMembers.get(0);
-            String memberName = payloadMember.getMemberName();
+            String payloadMemberName = payloadMember.getMemberName();
             writer.write("message.body = $L || message.body;",
-                    getInputValue(context, Location.PAYLOAD, "input." + memberName, payloadMember,
+                    getInputValue(context, Location.PAYLOAD, "input." + payloadMemberName, payloadMember,
                             model.expectShape(payloadMember.getTarget())));
-        } else if (!documentMembers.isEmpty()) {
-            // Write event document bindings if exist.
+        } else if (payloadShape instanceof StructureShape || payloadShape instanceof UnionShape) {
+            // handle implicit event payload by removing members with eventHeader trait.
+            List<MemberShape> headerMembers = event.getAllMembers().values().stream()
```

Comment:
This can be simplified a bit:

```
// Handle implicit event documents by removing event headers before serializing.
for (MemberShape memberShape : event.getMembers()) {
    if (memberShape.hasTrait(EventHeaderTrait.class)) {
        writer.write("delete input[$S]", memberShape.getMemberName());
    }
}
```

Reply:
Addressed in https://github.com/awslabs/smithy-typescript/pull/554/commits/f3e9ea325ee28184a5c2d145ca361a18a52e239e